File: algopatterns/dp/top_down_memo.cpp
// Pattern: Top-down DP with memoization
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

using State = string;
using Result = int;

unordered_map<State, Result> memo;

Result solve(State s) {
    // -- INSERT BASE CASE HERE --
    if (memo.count(s))
        return memo[s];
    // -- INSERT RECURSIVE CASE HERE --
    return memo[s] = {}; // placeholder
}

int main() {
    State initial_state;
    cout << solve(initial_state) << endl;
    return 0;
}

File: algopatterns/dp/top_down_memo.py
# Pattern: Top-down DP with manual memoization
memo = {}

def solve(state):
    # -- INSERT BASE CASE HERE --
    if state in memo:
        return memo[state]
    # -- INSERT RECURSIVE CASE HERE --
    # memo[state] = result
    pass

if __name__ == "__main__":
    initial_state = None
    print(solve(initial_state))
